addons/mass_mailing/models/mailing_contact.py

```diff
@@ -18,6 +19,14 @@ class MassMailingContactListRel(models.Model):
     list_id = fields.Many2one('mailing.list', string='Mailing List', ondelete='cascade', required=True)
     opt_out = fields.Boolean(string='Opt Out',
                              help='The contact has chosen not to receive mails anymore from this list', default=False)
+    opt_out_reason = fields.Selection(string='Opt Out Reason', selection=[
+        ('never_subscribed', 'I never subscribed to this list'),
+        ('changed_mind', 'I changed my mind'),
+        ('too_many_emails', 'I receive too many emails from this list'),
+        ('irrelevant_content', 'The content of these emails is not relevant to me'),
+        ('other', 'Other'),
+    ])
+    opt_out_custom = fields.Text(string='Opt Out Custom Reason')
```

Comment:
Not sure I would store it in a specific field. Simply posting it on the record message history should be sufficient.